### PR TITLE
[WIP] soundwire: move sdw_master_device_add to sdw_add_bus_master

### PR DESCRIPTION
**Pull request by bardliao**
Currently, we use a new sdw master API (sdw_master_device_add) to add sdw master devices. Some reviewers ask us to reuse the existing API (sdw_add_bus_master) to do so.
This PR is still WIP. However, could you give me some comments? I would like to know what do you think about this change.
Thanks.

### CODE REVIEW

File: drivers/soundwire/bus.c [hunk 1]
```diff
@@ -16,27 +16,47 @@
...
 {
 	struct sdw_master_prop *prop = NULL;
+	struct sdw_master_device *md;
 	int ret;
 
+	md = sdw_master_device_add(parent,
+                                   fwnode,
+                                   link_ops,
+                                   bus->link_id,
+                                   pdata);
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: I would have done the exact opposite. The main API is sdw_master_device_add, which calls the link .add callback that does all the inits and then you add the sdw_add_bus_master as the last stage of sdw_master_device_add().

In other words sdw_add_bus_master() is a no longer an externally visible API but an internal one.

Comment (bardliao) [reply drivers/soundwire/bus.c]: > I would have done the exact opposite. The main API is sdw_master_device_add, which calls the link .add callback that does all the inits and then you add the sdw_add_bus_master as the last stage of sdw_master_device_add().
> 
> In other words sdw_add_bus_master() is a no longer an externally visible API but an internal one.

I call sdw_add_bus_master() in sdw_intel_probe_controller() and sdw_add_bus_master() has almost the same parameters as sdw_master_device_add(). So as I can imagine that your proposal will be very similar to this PR, right?

File: drivers/soundwire/bus.c [hunk 2]
```diff
@@ -16,27 +16,47 @@
...
+	if (IS_ERR(md)) {
+		pr_err("Could not create link %d\n", bus->link_id);
+	}
```
Comment (lyakh) [inline drivers/soundwire/bus.c]: superfluous parentheses

File: drivers/soundwire/bus.c [hunk 3]
```diff
@@ -16,27 +16,47 @@
...
+	md->bus = bus;
+	bus->dev = &md->dev;
+
+	if (!bus) {
+		pr_err("SoundWire bus is null\n");
+		return ERR_PTR(-ENODEV);
+	}
```
Comment (lyakh) [inline drivers/soundwire/bus.c:34]: `bus` is a function input parameter. Do you really want to check it *after* you've called `sdw_master_device_add()`? Let alone after dereferencing it multiple times.

File: drivers/soundwire/intel_init.c [hunk 4]
```diff
@@ -288,6 +287,7 @@ static struct sdw_intel_ctx
 
 	ctx->num_slaves = num_slaves;
 	i = 0;
+
```
Comment (lyakh) [inline drivers/soundwire/intel_init.c]: doesn't belong to this patch

File: drivers/soundwire/qcom.c [hunk 5]
```diff
@@ -821,8 +821,10 @@ static int qcom_swrm_probe(struct platform_device *pdev)
 		goto err_clk;
 	}
 
-	ret = sdw_add_bus_master(&ctrl->bus);
-	if (ret) {
+	ctrl->md = sdw_add_bus_master(&ctrl->bus, dev, dev->fwnode,
+				      NULL, NULL);
```
Comment (lyakh) [inline drivers/soundwire/qcom.c]: this breaks bisection, you cannot do this separately from actually changing the function

**Comment by bardliao** [comment]
@plbossart @lyakh I have updated my PR. The major change is to add struct sdw_master_device md in struct sdw_bus {}. Could you review it again?

File: drivers/soundwire/bus.c [hunk 6]
```diff
@@ -16,11 +16,23 @@
  * Initializes the bus instance, read properties and create child
  * devices.
  */
+
 int sdw_add_bus_master(struct sdw_bus *bus)
 {
 	struct sdw_master_prop *prop = NULL;
 	int ret;
 
+	if (!bus) {
```
Comment (lyakh) [inline drivers/soundwire/bus.c:25]: but can it really be NULL? Can happen of course in principle, but are such cases really expected? So far I only see calls where NULL is impossible

File: include/linux/soundwire/sdw.h [hunk 7]
```diff
@@ -650,10 +650,7 @@ struct sdw_link_ops;
 struct sdw_master_device {
 	struct device dev;
 	struct sdw_bus *bus;
-	struct sdw_link_ops *link_ops;
-	int link_id;
 	bool pm_runtime_suspended;
-	void *pdata;
```
Comment (lyakh) [inline include/linux/soundwire/sdw.h:656]: I think it would make reviewing this patch easier if you could split it into several. The difficulty is, that this patch seems to be doing several things: (1) removing several fields out of `struct sdw_master_device` and switching over to using those from `struct sdw_bus`, adding them there if needed, (2) performing various functional changes. Maybe it could be split at least into: (1) move `.link_ops`, (2) move `.link_id`, (3) move `pdata`, (4) do the rest?

File: drivers/soundwire/bus.c (hunk 6 above)
Comment (paulstelian97) [reply drivers/soundwire/bus.c:25]: Do we want to just Oops if some bug makes it NULL?

**Comment by bardliao** [comment]
replaced by #2051

File: include/linux/soundwire/sdw.h (hunk 7 above)
Comment (bardliao) [reply include/linux/soundwire/sdw.h:656]: > I think it would make reviewing this patch easier if you could split it into several. The difficulty is, that this patch seems to be doing several things: (1) removing several fields out of `struct sdw_master_device` and switching over to using those from `struct sdw_bus`, adding them there if needed, (2) performing various functional changes. Maybe it could be split at least into: (1) move `.link_ops`, (2) move `.link_id`, (3) move `pdata`, (4) do the rest?

@lyakh #2051 Should make it easier :)

File: drivers/soundwire/bus.c (hunk 6 above)
Comment (bardliao) [reply drivers/soundwire/bus.c:25]: > Do we want to just Oops if some bug makes it NULL?

Good idea. Thanks @lyakh @paulstelian97

Comment (plbossart) [reply drivers/soundwire/bus.c:25]: Given that the bus structure is allocated by something else, I find it's a safe coding practice to check that it's not NULL...
